sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlReplace.java

```diff
@@ -0,0 +1,133 @@
+
+  private final Granularity partitionedBy;
+  private final String partitionedByStringForUnparse;
+
+  private final List<String> replaceTimeChunks;
+
+  /**
+   * While partitionedBy and partitionedByStringForUnparse can be null as arguments to the constructor, this is
+   * disallowed (semantically) and the constructor performs checks to ensure that. This helps in producing friendly
+   * errors when the PARTITIONED BY custom clause is not present, and keeps its error separate from JavaCC/Calcite's
+   * custom errors which can be cryptic when someone accidentally forgets to explicitly specify the PARTITIONED BY clause
+   */
+  public DruidSqlReplace(
+      @Nonnull SqlInsert insertNode,
+      @Nullable Granularity partitionedBy,
+      @Nullable String partitionedByStringForUnparse,
+      @Nonnull List<String> replaceTimeChunks
+  ) throws ParseException
+  {
+    super(
+        insertNode.getParserPosition(),
+        (SqlNodeList) insertNode.getOperandList().get(0), // No better getter to extract this
+        insertNode.getTargetTable(),
+        insertNode.getSource(),
+        insertNode.getTargetColumnList()
+    );
+    if (partitionedBy == null) {
+      throw new ParseException("REPLACE statements must specify PARTITIONED BY clause explictly");
+    }
+    this.partitionedBy = partitionedBy;
+
+    Preconditions.checkNotNull(partitionedByStringForUnparse);
+    this.partitionedByStringForUnparse = partitionedByStringForUnparse;
+
+    this.replaceTimeChunks = replaceTimeChunks;
+  }
+
+  public List<String> getReplaceTimeChunks()
+  {
+    return replaceTimeChunks;
+  }
+
+  public Granularity getPartitionedBy()
+  {
+    return partitionedBy;
+  }
+
+  @Nonnull
+  @Override
+  public SqlOperator getOperator()
+  {
+    return OPERATOR;
+  }
+
+  @Override
+  public void unparse(SqlWriter writer, int leftPrec, int rightPrec)
+  {
+    writer.startList(SqlWriter.FrameTypeEnum.SELECT);
+    writer.sep("REPLACE INTO");
+    final int opLeft = getOperator().getLeftPrec();
+    final int opRight = getOperator().getRightPrec();
+    getTargetTable().unparse(writer, opLeft, opRight);
+
+    writer.keyword("FOR");
+    final SqlWriter.Frame frame = writer.startList("(", ")");
+    List<String> replaceTimeChunks = getReplaceTimeChunks();
+    for (String replaceTimeChunk : replaceTimeChunks) {
+      writer.sep(","); // sep() takes care of not printing the first separator
+      writer.literal(unparseTimeChunk(replaceTimeChunk));
+    }
+    writer.endList(frame);
+
+    if (getTargetColumnList() != null) {
+      getTargetColumnList().unparse(writer, opLeft, opRight);
+    }
+    writer.newlineAndIndent();
+    getSource().unparse(writer, 0, 0);
+    writer.keyword("PARTITIONED BY");
+    writer.keyword(partitionedByStringForUnparse);
+  }
+
+  private String unparseTimeChunk(String timeChunkString)
+  {
+    if ("all".equals(timeChunkString)) {
+      return "ALL TIME";
```

Comment:
The javadocs for unparse doesn't mention this, only that it has to be a SQL representation of the node. I also didn't understand your comment, is unparsing "ALL TIME", which is the same form that would have been parsed to "all" not acceptable?

Reply:
I meant something like if a user enters
``` insert INTO foo select * from bar partitioned BY all```
we should show the unparse string as
```INSERT INTO foo SELECT * FROM bar PARTITIONED BY ALL```
rather than
```INSERT INTO foo SELECT * FROM bar PARTITIONED BY ALL TIME```